### PR TITLE
check slice_min/max(order_by) and slice_sample(weight_by=)

### LINKED ISSUE #5922: slice_max returns empty rows for grouped_df when ordered by environment variable
`slice_max` returns empty data.frame only when the data.frame is grouped and a variable from `.GlobalEnv` is used for ordering. Reprex below.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
.df <- data.frame(
  A=rep(letters[1:2], each=6),
  B=rep(letters[1:3], each=4),
  C=1:12
)

V <- .df$C
# slice_max with environment var fails
.df %>% group_by(A) %>% slice_max(V, n=2)
#> # A tibble: 0 x 3
#> # Groups:   A [0]
#> # … with 3 variables: A <chr>, B <chr>, C <int>

# slice_max with data var works
.df %>% group_by(A) %>% slice_max(C, n=2)
#> # A tibble: 4 x 3
#> # Groups:   A [2]
#>   A     B         C
#>   <chr> <chr> <int>
#> 1 a     b         6
#> 2 a     b         5
#> 3 b     c        12
#> 4 b     c        11

# slice_max with ungrouped data.frame works
.df %>% slice_max(V, n=2)
#>   A B  C
#> 1 b c 12
#> 2 b c 11

# slice_min with environment var works
.df %>% group_by(A) %>% slice_min(V, n=2)
#> # A tibble: 4 x 3
#> # Groups:   A [2]
#>   A     B         C
#>   <chr> <chr> <int>
#> 1 a     a         1
#> 2 a     a         2
#> 3 b     b         7
#> 4 b     b         8

sessionInfo()
#> R version 4.0.3 (2020-10-10)
#> Platform: x86_64-apple-darwin17.0 (64-bit)
#> Running under: macOS High Sierra 10.13.6
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
#> LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] dplyr_1.0.5
#> 
#> loaded via a namespace (and not attached):
#>  [1] rstudioapi_0.13   knitr_1.31        magrittr_2.0.1    tidyselect_1.1.0 
#>  [5] debugme_1.1.0     R6_2.5.0          rlang_0.4.10      fansi_0.4.2      
#>  [9] stringr_1.4.0     styler_1.4.1      highr_0.8         tools_4.0.3      
#> [13] xfun_0.22         utf8_1.2.1        cli_2.4.0         DBI_1.1.1        
#> [17] withr_2.4.1       htmltools_0.5.1.1 ellipsis_0.3.1    assertthat_0.2.1 
#> [21] yaml_2.2.1        digest_0.6.27     tibble_3.1.0      lifecycle_1.0.0  
#> [25] crayon_1.4.1      purrr_0.3.4       vctrs_0.3.7       fs_1.5.0         
#> [29] glue_1.4.2        evaluate_0.14     rmarkdown_2.7     reprex_2.0.0     
#> [33] stringi_1.5.3     compiler_4.0.3    pillar_1.5.1      generics_0.1.0   
#> [37] backports_1.2.1   pkgconfig_2.0.3
```

<sup>Created on 2021-06-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #5922

``` r
library(dplyr, warn.conflicts = FALSE)
df <- data.frame(x = 1:10)
slice_min(df, 1:6)
#> Error: `slice_min(order_by=)` must have size 10, not size 6.
slice_max(df, 1:6)
#> Error: `slice_max(order_by=)` must have size 10, not size 6.
slice_sample(df, n = 2, weight_by = 1:6)
#> Error: `slice_sample(weight_by=)` must have size 10, not size 6.
```

<sup>Created on 2021-06-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### CODE REVIEW

File: R/slice.R [hunk 1]
```diff
@@ -185,7 +192,7 @@ slice_min.data.frame <- function(.data, order_by, ..., n, prop, with_ties = TRUE
       prop = function(x, n) head(order(x), size$prop * n)
     )
   }
-  slice(.data, idx({{ order_by }}, dplyr::n()))
+  slice(.data, check_slice_by(idx, "slice_min(order_by=)", {{ order_by }}, dplyr::n()))
```
Comment (romainfrancois) [inline R/slice.R:195]: Maybe it's clearer to just add the `vec_assert()` in the various forms of `idx()` instead of having this additional function ?

Comment (hadley) [reply R/slice.R:195]: It's kind of marginal either way here, but there are 3 cases which is where we normally recommend using a function.

File: tests/testthat/_snaps/slice.md [hunk 2]
```diff
@@ -1,3 +1,21 @@
+# slice_min/max() check size of `order_by=` (#5922)
+
+    Code
+      slice_min(data.frame(x = 1:10), 1:6)
+    Error <vctrs_error_assert_size>
+      `slice_min(order_by=)` must have size 10, not size 6.
```
[the remark this answers is not in the document]
Comment (hadley) [reply tests/testthat/_snaps/slice.md:6]: I think our usual convention is "`order_by` must have size ...".  We're planning on some rlang changes to make it easier to systematically include the function name in the erorr (e.g. "Error in slice_min(): `order_by` must have size...") so don't worry about that here.

**Review by hadley — APPROVED** [review]

File: R/slice.R [hunk 3]
```diff
@@ -153,6 +153,13 @@ slice_tail.data.frame <- function(.data, ..., n, prop) {
   slice(.data, idx(dplyr::n()))
 }
 
+check_slice_by <- function(idx, order_by, n) {
+  if (!is.null(order_by)) {
+    vec_assert(order_by, size = n)
```
Comment (DavisVaughan) [inline R/slice.R:158]:
```suggestion
    vec_assert(order_by, size = n, arg = "order_by")
```

Since I think users could see this?

Comment (DavisVaughan) [reply R/slice.R:158]: Although I guess `vec_assert()` will do that for you...

**Comment by romainfrancois** [comment]
I believe this form: 

```r
slice(.data, {
    n <- dplyr::n()
    x <- vec_assert({{ order_by }}, size = n, arg = "order_by")
    idx(x, n)
  })
```

to be slightly more readable than having an extra function that eventually just calls `idx`. The problem this has is `order_by` and `weight_by` have to be evaluated lazily "inside" the `slice()` call. 

This also eliminates the need to name the function: 

```r
slice(.data, check_slice_by(idx, {{ order_by }}, dplyr::n(), arg = "order_by"))
```

`check_slice_by()` probably wasn't the good name, and this form feels like it hides what's happening.

**Review by DavisVaughan — APPROVED** [review]

**Pull request merged**